### PR TITLE
OptimizedLinear updates

### PR DESCRIPTION
**Pull request by jeffra**
This is a refresh of of `OptimizedLinear` with the following features to improve performance and usability:
 * More efficient sharing of base weights using `all_gather_into_tensor`
 * Flattened sharded weights
 * Selectively offload frozen weights to cpu
 * `deepspeed.linear.Init` that allows injecting OptimizedLinear during model construction (similar to zero.Init)
 * Support for load state dict directly in OptimizedLinear, this allows loading HF model weights correctly into sharded params
 * Various bug fixes for the LoRA implementation introduced previously
 * Several new unit tests
 
Builds on-top of @RezaYazdaniAminabadi's previous FP8 updates (#5764) to support dense model fp8 quantization. 

Example usage of this to fine-tune llama-3.1-405B on a single node: https://github.com/Snowflake-Labs/snowflake-arctic/tree/main/training/llama3.1

### CODE REVIEW

File: deepspeed/linear/optimized_linear.py [hunk 1]
```diff
@@ -85,50 +85,69 @@ def __init__(self,
...
+
+        self.weight.requires_grad = False
 
-        self.base_weight.requires_grad = False
+        # Mark base weight to prevent broadcast and ensure proper offload behavior
+        self.weight.ds_optim_param = True
 
         # Use RS lora for now.
-        self.lora_scaling_factor = self.lora_config.lora_alpha / math.sqrt(self.lora_config.lora_r)
+        self.lora_scaling_factor = self.lora_config.lora_alpha / self.lora_config.lora_r
```
Comment (winglian) [inline deepspeed/linear/optimized_linear.py:111]: You may want to remove `# Use RS lora for now.` comment as the `math.sqrt(...)` that was removed is the rs-lora technique.

Comment (sfc-gh-jrasley) [reply deepspeed/linear/optimized_linear.py:111]: thanks @winglian! good catch, this comment is now misleading :) This PR needs some general clean-up after I get #5764 merged, will address it then.

**Comment by jeffra** [comment]
`nv-accelerate-v100` results on H100 (torch 2.4 + cu12.2):
image

`nv-torch-latest-v100` results on H100 (torch 2.4 + cu12.2):
in-progress...

**Comment by jeffra** [comment]
I'm able to get both the `nv-accelerate-v100` and `nv-torch-latest-v100` workflows to pass with this branch on my local H100 node. /cc @tjruwase @HeyangQin @loadams. Okay to force merge?

**Review by HeyangQin — APPROVED** [review]

**Comment by loadams** [comment]
> I'm able to get both the `nv-accelerate-v100` and `nv-torch-latest-v100` workflows to pass with this branch on my local H100 node (see previous comment). /cc @tjruwase @HeyangQin @loadams. Okay to force merge?

I believe I've fixed our runners @jeffra - I'll monitor it today to be sure it gets merged.

**Pull request merged**